SimpleBlocksIT.testAddBlockWhileDeletingIndices failing

### CI Link

https://gradle-enterprise.elastic.co/s/uz5p3xwiwfsqq

### Repro line

./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.blocks.SimpleBlocksIT.testAddBlockWhileDeletingIndices" -Dtests.seed=18DD19966E2CF499 -Dtests.locale=dyo-SN -Dtests.timezone=Asia/Ust-Nera -Druntime.java=23

### Does it reproduce?

Didn't try

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

Likely introduced by PR https://github.com/elastic/elasticsearch/pull/115341 

```
REPRODUCE WITH: ./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.blocks.SimpleBlocksIT.testAddBlockWhileDeletingIndices" -Dtests.seed=18DD19966E2CF499 -Dtests.locale=dyo-SN -Dtests.timezone=Asia/Ust-Nera -Druntime.java=23

SimpleBlocksIT > testAddBlockWhileDeletingIndices FAILED
    java.lang.AssertionError: [org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener/ChannelActionListener{TaskTransportChannel{task=160}{DirectResponseChannel{req=91}{indices:admin/block/add[s][p]}}}/org.elasticsearch.action.support.replication.TransportReplicationAction$$Lambda/0x00007f63e3a6ebc8@72916952] org.elasticsearch.ElasticsearchException: executed already
        at __randomizedtesting.SeedInfo.seed([18DD19966E2CF499]:0)
        at org.elasticsearch.action.ActionListener$3.assertFirstRun(ActionListener.java:393)
        at org.elasticsearch.action.ActionListener$3.onFailure(ActionListener.java:409)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.onFailure(TransportReplicationAction.java:553)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.handleException(TransportReplicationAction.java:547)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.runWithPrimaryShardReference(TransportReplicationAction.java:541)
        at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.lambda$doRun$0(TransportReplicationAction.java:443)
        at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:257)
        at org.elasticsearch.index.shard.IndexShard.lambda$wrapPrimaryOperationPermitListener$38(IndexShard.java:3585)
        at org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:219)
        at org.elasticsearch.index.shard.IndexShard.lambda$asyncBlockOperations$39(IndexShard.java:3597)
        at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:257)
        at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:400)
        at org.elasticsearch.index.shard.IndexShardOperationPermits$1.doRun(IndexShardOperationPermits.java:119)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1575)
WARNING: A terminally deprecated method in java.lang.System has been called
WARNING: System::setSecurityManager has been called by org.gradle.api.internal.tasks.testing.worker.TestWorker (file:/opt/buildkite-agent/.gradle/wrapper/dists/gradle-8.10.2-all/7iv73wktx1xtkvlq19urqw1wm/gradle-8.10.2/lib/plugins/gradle-testing-base-infrastructure-8.10.2.jar)
WARNING: Please consider reporting this to the maintainers of org.gradle.api.internal.tasks.testing.worker.TestWorker
WARNING: System::setSecurityManager will be removed in a future release
```

--- a/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
+++ b/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
@@ -115,28 +115,25 @@ public class ReplicationOperation<
             );
             resultListener.onResponse(primaryResult);
         }, resultListener::onFailure))) {
-            final String activeShardCountFailure = checkActiveShardCount();
-            final ShardRouting primaryRouting = primary.routingEntry();
-            final ShardId primaryId = primaryRouting.shardId();
-            if (activeShardCountFailure != null) {
-                pendingActionsListener.acquire()
-                    .onFailure(
-                        new UnavailableShardsException(
-                            primaryId,
-                            "{} Timeout: [{}], request: [{}]",
-                            activeShardCountFailure,
-                            request.timeout(),
-                            request
-                        )
+            ActionListener.run(pendingActionsListener.acquire(), (primaryCoordinationListener) -> { // triggered when we finish coordination
+                final String activeShardCountFailure = checkActiveShardCount();
+                final ShardRouting primaryRouting = primary.routingEntry();
+                final ShardId primaryId = primaryRouting.shardId();
+                if (activeShardCountFailure != null) {
+                    throw new UnavailableShardsException(
+                        primaryId,
+                        "{} Timeout: [{}], request: [{}]",
+                        activeShardCountFailure,
+                        request.timeout(),
+                        request
                     );
-                return;
-            }
+                }
 
-            totalShards.incrementAndGet();
-            var primaryCoordinationPendingActionListener = pendingActionsListener.acquire(); // triggered when we finish all coordination
-            primary.perform(request, primaryCoordinationPendingActionListener.delegateFailureAndWrap((l, primaryResult) -> {
-                handlePrimaryResult(primaryResult, l, pendingActionsListener);
-            }));
+                totalShards.incrementAndGet();
+                primary.perform(request, primaryCoordinationListener.delegateFailureAndWrap((l, primaryResult) -> {
+                    handlePrimaryResult(primaryResult, l, pendingActionsListener);
+                }));
+            });
         }
     }
 
@@ -153,24 +150,25 @@ public class ReplicationOperation<
             }
             final ReplicationGroup replicationGroup = primary.getReplicationGroup();
 
-            var primaryOperationPendingActionListener = pendingActionsListener.acquire();
-            replicasProxy.onPrimaryOperationComplete(
-                replicaRequest,
-                replicationGroup.getRoutingTable(),
-                ActionListener.wrap(ignored -> primaryOperationPendingActionListener.onResponse(null), exception -> {
-                    totalShards.incrementAndGet();
-                    shardReplicaFailures.add(
-                        new ReplicationResponse.ShardInfo.Failure(
-                            primary.routingEntry().shardId(),
-                            null,
-                            exception,
-                            ExceptionsHelper.status(exception),
-                            false
-                        )
-                    );
-                    primaryOperationPendingActionListener.onResponse(null);
-                })
-            );
+            ActionListener.run(pendingActionsListener.acquire(), primaryOperationPendingActionListener -> {
+                replicasProxy.onPrimaryOperationComplete(
+                    replicaRequest,
+                    replicationGroup.getRoutingTable(),
+                    ActionListener.wrap(ignored -> primaryOperationPendingActionListener.onResponse(null), exception -> {
+                        totalShards.incrementAndGet();
+                        shardReplicaFailures.add(
+                            new ReplicationResponse.ShardInfo.Failure(
+                                primary.routingEntry().shardId(),
+                                null,
+                                exception,
+                                ExceptionsHelper.status(exception),
+                                false
+                            )
+                        );
+                        primaryOperationPendingActionListener.onResponse(null);
+                    })
+                );
+            });
 
             // we have to get the replication group after successfully indexing into the primary in order to honour recovery semantics.
             // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
@@ -234,13 +232,14 @@ public class ReplicationOperation<
     ) {
         // if inSyncAllocationIds contains allocation ids of shards that don't exist in RoutingTable, mark copies as stale
         for (String allocationId : replicationGroup.getUnavailableInSyncShards()) {
-            var staleCopyPendingActionListener = pendingActionsListener.acquire();
-            replicasProxy.markShardCopyAsStaleIfNeeded(
-                replicaRequest.shardId(),
-                allocationId,
-                primaryTerm,
-                staleCopyPendingActionListener.delegateResponse((l, e) -> onNoLongerPrimary(e, l))
-            );
+            ActionListener.run(pendingActionsListener.acquire(), (staleCopyPendingActionListener) -> {
+                replicasProxy.markShardCopyAsStaleIfNeeded(
+                    replicaRequest.shardId(),
+                    allocationId,
+                    primaryTerm,
+                    staleCopyPendingActionListener.delegateResponse((l, e) -> onNoLongerPrimary(e, l))
+                );
+            });
         }
     }
 
@@ -285,13 +284,17 @@ public class ReplicationOperation<
             logger.trace("[{}] sending op [{}] to replica {} for request [{}]", shard.shardId(), opType, shard, replicaRequest);
         }
         totalShards.incrementAndGet();
-        var replicationPendingActionListener = pendingActionsListener.acquire();
-        ActionListener.run(replicationPendingActionListener, (listener) -> {
+        ActionListener.run(pendingActionsListener.acquire(), (replicationPendingActionListener) -> {
             final ActionListener<ReplicaResponse> replicationListener = new ActionListener<>() {
                 @Override
                 public void onResponse(ReplicaResponse response) {
                     successfulShards.incrementAndGet();
-                    updateCheckPoints(shard, response::localCheckpoint, response::globalCheckpoint, () -> listener.onResponse(null));
+                    updateCheckPoints(
+                        shard,
+                        response::localCheckpoint,
+                        response::globalCheckpoint,
+                        () -> replicationPendingActionListener.onResponse(null)
+                    );
                 }
 
                 @Override
@@ -325,7 +328,7 @@ public class ReplicationOperation<
                         primaryTerm,
                         message,
                         replicaException,
-                        listener.delegateResponse((l, e) -> onNoLongerPrimary(e, l))
+                        replicationPendingActionListener.delegateResponse((l, e) -> onNoLongerPrimary(e, l))
                     );
                 }
 

